support/agent/deal_client_agent.go

```diff
@@ -0,0 +1,206 @@
+	MinStoragePrice  abi.TokenAmount // minimum price per epoch a client will pay for storage (may be zero)
+	MaxStoragePrice  abi.TokenAmount // maximum price per epoch a client will pay for storage
+	MinMarketBalance abi.TokenAmount // balance below which client will top up funds in market actor
+	MaxMarketBalance abi.TokenAmount // balance to which client will top up funds in market actor
+}
+
+type DealClientAgent struct {
+	DealCount int
+
+	account    address.Address
+	config     DealClientConfig
+	dealEvents *RateIterator
+	rnd        *rand.Rand
+
+	// tracks funds expected to be locked for client deal payment
+	expectedMarketBalance abi.TokenAmount
+}
+
+func AddDealClientsForAccounts(s SimState, accounts []address.Address, seed int64, config DealClientConfig) []*DealClientAgent {
+	rnd := rand.New(rand.NewSource(seed))
+	var agents []*DealClientAgent
+	for _, account := range accounts {
+		agent := NewDealClientAgent(account, rnd.Int63(), config)
+		agents = append(agents, agent)
+		s.AddAgent(agent)
+	}
+	return agents
+}
+
+func NewDealClientAgent(account address.Address, seed int64, config DealClientConfig) *DealClientAgent {
+	rnd := rand.New(rand.NewSource(seed))
+	return &DealClientAgent{
+		account:               account,
+		config:                config,
+		rnd:                   rnd,
+		expectedMarketBalance: big.Zero(),
+		dealEvents:            NewRateIterator(config.DealRate, rnd.Int63()),
+	}
+}
+
+func (dca *DealClientAgent) Tick(s SimState) ([]message, error) {
+	var providers []DealProvider
+
+	// aggregate all deals into one message
+	if err := dca.dealEvents.Tick(func() error {
+		provider := s.ChooseDealProvider()
+		providers = append(providers, provider)
+		if err := dca.createDeal(s, provider); err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	// add message to update market balance if necessary
+	messages := dca.updateMarketBalance()
+
+	return messages, nil
+}
+
+func (dca *DealClientAgent) updateMarketBalance() []message {
+	if dca.expectedMarketBalance.GreaterThanEqual(dca.config.MinMarketBalance) {
+		return []message{}
+	}
+
+	balanceToAdd := big.Sub(dca.config.MaxMarketBalance, dca.expectedMarketBalance)
+
+	return []message{{
+		From:   dca.account,
+		To:     builtin.StorageMarketActorAddr,
+		Value:  balanceToAdd,
+		Method: builtin.MethodsMarket.AddBalance,
+		Params: &dca.account,
+		ReturnHandler: func(v SimState, msg message, ret cbor.Marshaler) error {
+			dca.expectedMarketBalance = dca.config.MaxMarketBalance
+			return nil
+		},
+	}}
+}
+
+// Create a proposal
+// Return false if provider if deal can't be performed because client or provider lacks funds
+func (dca *DealClientAgent) createDeal(s SimState, provider DealProvider) error {
+	pieceCid, err := dca.generatePieceCID()
+	if err != nil {
+		return err
+	}
+
+	pieceSize := dca.config.MinPieceSize +
+		uint64(dca.rnd.Int63n(int64(dca.config.MaxPieceSize-dca.config.MinPieceSize)))
```

Comment:
The goal is to assume a uniform distribution over some range of file sizes and then compute the padded piece size from that. This will give an uneven distribution of padded piece sizes which should be somewhat realistic while also covering the range.